README.md
# Parhelion

A simple Discord bot that can play music, solve math questions, and send reminders.

![demo](images/demo.png)

## how to try this bot
There are two ways you can add this bot to Discord servers where you have "manage server" permissions:

1. By adding my instance of the bot to your server by clicking [this link](https://discordapp.com/api/oauth2/authorize?scope=bot&client_id=836071320328077332&permissions=3595328).
2. By making a fork of this repl and creating your own instance of the bot (see instructions below).

After adding this bot, you can see all the commands by entering `;help`.

## how to create your own instance of this bot
1. Create a Discord bot account ([see this guide](https://www.freecodecamp.org/news/create-a-discord-bot-with-python/)), except don't create an `.env` file because Replit doesn't use those anymore. Instead, follow [this guide](https://docs.replit.com/repls/secrets-environment-variables) to set a Replit environment variable called `DISCORD_BOT_SECRET_TOKEN`.
2. To keep the bot running continuously without Replit's premium "always on" option, you can use a free UptimeRobot account as explained in the first guide linked above.
3. Replit isn't designed to host bots, so you will want to change hosts soon if you continue using the bot.

## guides
Here are some guides to making Discord bots that I found helpful:
* https://ritza.co/showcase/repl.it/building-a-discord-bot-with-python-and-repl-it.html
* https://www.freecodecamp.org/news/create-a-discord-bot-with-python/
* https://discordpy.readthedocs.io/en/latest/index.html
* https://github.com/Rapptz/discord.py/tree/v1.7.1/examples
* https://discord.com/developers/docs/intro

## commands
Here are some of this bot's commands. This image will not always be kept up-to-date.

![help demo](images/help_demo.png)


CHANGELOG.md
## Below are git log messages from before I created this file (CHANGELOG.md).

Date:Fri Jun 4 16:11:49 2021 +0000  
    Improve join command error messages

Date: Fri Jun 4 07:30:13 2021 +0000  
    Improve docstrings

Date: Fri Jun 4 05:54:30 2021 +0000  
    Fix continue_reminders

Date: Fri Jun 4 02:32:17 2021 +0000  
    Give more help on mention

Date: Fri Jun 4 02:24:56 2021 +0000  
    Add exe alias to py command

Date: Fri Jun 4 02:24:14 2021 +0000  
    Simplify reminder message sanitization

Date: Fri Jun 4 02:00:52 2021 +0000  
    Remove unnecessary TODO

Date: Fri Jun 4 01:59:08 2021 +0000  
    Improve owner's command  context saving

Date: Thu Jun 3 17:11:03 2021 +0000  
    Replace unicode bullet codes with literals

Date: Thu Jun 3 17:05:17 2021 +0000  
    Improve docstring

Date: Thu Jun 3 07:37:55 2021 +0000  
    Remove unnecessary except

Date: Thu Jun 3 07:34:41 2021 +0000  
    Log reminder deletion

Date: Thu Jun 3 05:17:05 2021 +0000  
    Remove tts variable

Date: Thu Jun 3 04:49:26 2021 +0000  
    Remove get_shortest_prefix function

Date: Thu Jun 3 00:47:34 2021 +0000  
    Update TODOs

Date: Thu Jun 3 00:45:23 2021 +0000  
    Add docstrings and a comment

Date: Thu Jun 3 00:11:08 2021 +0000  
    Create prefix command

Date: Thu Jun 3 00:03:40 2021 +0000  
    Create dev_settings object, use when_mentioned_or

Date: Wed Jun 2 20:11:31 2021 +0000  
    Update TODO.md

Date: Wed Jun 2 20:08:40 2021 +0000  
    Rename command exec to py, add aliases to several commands

Date: Wed Jun 2 19:09:38 2021 +0000  
    Remove reverse command

Date: Wed Jun 2 19:08:53 2021 +0000  
    Remove roll command

Date: Wed Jun 2 19:04:49 2021 +0000  
    Add space

Date: Wed Jun 2 19:00:01 2021 +0000  
    Allow multiple extensions per extension command call

Date: Wed Jun 2 18:42:13 2021 +0000  
    Create list-exts command

Date: Wed Jun 2 18:33:26 2021 +0000  
    React to successful exec with check mark

Date: Wed Jun 2 07:57:02 2021 +0000  
    Add ping prefix and help with prefixes on ping

Date: Tue Jun 1 20:05:10 2021 +0000  
    Add aliases to exec and eval

Date: Tue Jun 1 20:02:51 2021 +0000  
    Improve error message

Date: Tue Jun 1 20:02:19 2021 +0000  
    Improve comments

Date: Tue Jun 1 19:59:45 2021 +0000  
    Add getsource alias to inspect

Date: Tue Jun 1 19:57:02 2021 +0000  
    Add users alias to about

Date: Tue Jun 1 19:56:16 2021 +0000  
    Bring back the dev-about command

Date: Tue Jun 1 19:34:25 2021 +0000  
    Remove command use counting    
    Use PostgreSQL to log all command use in the future?  

Date: Tue Jun 1 19:29:31 2021 +0000  
    Add spaces under each field in about

Date: Tue Jun 1 08:26:07 2021 +0000  
    Create server-info command

Date: Tue Jun 1 08:04:39 2021 +0000  
    Count command uses

Date: Tue Jun 1 07:49:01 2021 +0000  
    Improve about command

Date: Tue Jun 1 04:01:24 2021 +0000  
    Improve docstrings

Date: Mon May 31 21:02:17 2021 +0000  
    Move dev_mail to common

Date: Mon May 31 20:59:58 2021 +0000  
    Remove unnecessary imports

Date: Mon May 31 20:58:32 2021 +0000  
    Custom math error message for

Date: Mon May 31 20:49:49 2021 +0000  
    Update TODO.md

Date: Mon May 31 19:55:52 2021 +0000  
    Improve continue_reminder

Date: Mon May 31 19:51:26 2021 +0000  
    Create language params in remove_backticks

Date: Mon May 31 19:13:46 2021 +0000  
    Change reloadall to reload-all

Date: Mon May 31 19:12:35 2021 +0000  
    Add py and python aliases

Date: Mon May 31 06:15:39 2021 +0000  
    Create eval_reminder function

Date: Mon May 31 06:15:25 2021 +0000  
    Update TODO.md

Date: Mon May 31 05:01:31 2021 +0000  
    Return type hint

Date: Mon May 31 05:00:21 2021 +0000  
    Put cooldown on dev-about

Date: Mon May 31 04:54:40 2021 +0000  
    Move dev-about command to Other cog

Date: Mon May 31 04:53:02 2021 +0000  
    Improve comments

Date: Mon May 31 04:51:15 2021 +0000  
    Move common functions

Date: Mon May 31 04:49:03 2021 +0000  
    Use the mathjs API for the calc command

Date: Sun May 30 21:42:58 2021 +0000  
    Comment about eval and exec security

Date: Sun May 30 21:11:36 2021 +0000  
    Create dev-about command

Date: Sun May 30 21:10:17 2021 +0000  
    Escape double quotes, delete glitched reminders

Date: Sat May 29 18:04:28 2021 +0000  
    Fix continue_reminder

Date: Sat May 29 07:01:01 2021 +0000  
    Use the forismatic API for the quote command

Date: Sat May 29 05:48:44 2021 +0000  
    Create the close command

Date: Fri May 28 23:37:15 2021 +0000  
    Fix list copy in reloadall

Date: Fri May 28 19:37:19 2021 +0000  
    Reduce sends from reloadall

Date: Fri May 28 19:26:55 2021 +0000  
    Improve del-r error handling

Date: Fri May 28 19:23:09 2021 +0000  
    Improve error handling for the choose command

Date: Fri May 28 18:55:11 2021 +0000  
    Customize error handling for the reload command

Date: Fri May 28 18:30:42 2021 +0000  
    Create reloadall command

Date: Fri May 28 18:25:56 2021 +0000  
    Add aliases to the about command

Date: Fri May 28 07:02:30 2021 +0000  
    Add spaces to about's embed

Date: Thu May 27 20:23:15 2021 +0000  
    Improve the about command

Date: Thu May 27 18:04:52 2021 +0000  
    Improve organization

Date: Thu May 27 17:45:28 2021 +0000  
    Allow the r command to skip commands

Date: Thu May 27 17:30:16 2021 +0000  
    Remove reminders.txt from .gitignore

Date: Thu May 27 04:38:11 2021 +0000  
    Use Replit's database to save reminders

Date: Wed May 26 19:43:33 2021 +0000  
    Change tabs to spaces

Date: Wed May 26 19:25:39 2021 +0000  
    Reorganize README.md

Date: Wed May 26 18:45:43 2021 +0000  
    Update TODO.md

Date: Wed May 26 18:38:37 2021 +0000  
    Add favicon.ico to prevent 404 errors

Date:   Wed May 26 07:33:33 2021 +0000  
    Extend peat_command to be able to repeat multiple commands

Date: Wed May 26 07:17:04 2021 +0000  
    Handle errors from the choose command


other.py
import os
import platform
import inspect
import discord
from discord.ext import commands


async def dev_mail(bot, message: str, use_embed: bool = True, embed_title: str = 'dev mail'):
	user = await bot.fetch_user(int(os.environ['MY_USER_ID']))
	if use_embed:
		embed = discord.Embed(title=embed_title, description=message)
		await user.send(embed=embed)
	else:
		await user.send(message)


class Other(commands.Cog):
	def __init__(self, bot):
		self.bot = bot


	@commands.command(hidden=True)
	@commands.cooldown(3, 15)
	async def hhelp(self, ctx):
		'''Shows help for all the hidden commands'''
		hidden_commands = []
		for cmd in self.bot.commands:
			if cmd.hidden:
				hidden_commands.append(cmd)

		# Alphabetize.
		hidden_commands = sorted(hidden_commands, key=lambda x: x.name)

		# Get column width.
		hidden_names = [x.name for x in hidden_commands]
		width = len(max(hidden_names, key=len))

		message = 'Hidden Commands:'
		for cmd in hidden_commands:
			message += f'\n  {cmd.name:<{width}} {cmd.short_doc}'
			if len(cmd.checks):
				message += ' (bot owner only)'
		message += '\n\n Type ;help command for more info on a command.'

		await ctx.send(f'```{message}```')


	@commands.command(hidden=True)
	@commands.cooldown(3, 15)
	async def echo(self, ctx, *, message: str):
		'''Repeats a message'''
		await ctx.send(message)


	@commands.command(hidden=True)
	@commands.cooldown(3, 15)
	async def ping(self, ctx):
		'''Pings the server'''
		await ctx.send(f'Pong! It took {round(self.bot.latency, 2)} ms.')


	@commands.command(aliases=['info', 'stats', 'invite'])
	@commands.cooldown(3, 15)
	async def about(self, ctx):
		'''Shows general info about this bot'''
		embed = discord.Embed(
			title='About me',
			description= f'''
				Created by Chris Wheeler
				with Python {platform.python_version()} and [discord.py](https://discordpy.readthedocs.io/en/latest/)

				Currently in {len(self.bot.guilds)} servers.
				Invite link [here](https://discordapp.com/api/oauth2/authorize?scope=bot&client_id=836071320328077332&permissions=3300352)
				Source code [here](https://replit.com/@wheelercj/simple-Discord-bot)
			'''
		)
		
		await ctx.send(embed=embed)


	@commands.command(name='inspect', aliases=['source', 'src'])
	@commands.cooldown(3, 15)
	async def _inspect(self, ctx, *, command: str):
		'''Shows the source code of a command'''
		try:
			cmds = {cmd.name: cmd for cmd in self.bot.commands}
			if command not in cmds.keys():
				raise NameError(f'Command {command} not found.')
			source = str(inspect.getsource(cmds[command].callback))
			await ctx.send(f'```py\n{source}```')
		except NameError as e:
			await ctx.send(e)
		except KeyError as e:
			await ctx.send(e)


	@commands.command()
	@commands.cooldown(3, 15)
	async def calc(self, ctx, *, string: str):
		'''Evaluates a math expression
		
		Uses a limited version of Python's eval function.'''
		try:
			# The eval function can do just about anything by default, so a
			# lot of its features have to be removed for security. For more
			# info, see https://realpython.com/python-eval-function/#minimizing-the-security-issues-of-eval 
			allowed_names = {}
			code = compile(string, '<string>', 'eval')
			for name in code.co_names:
				if name not in allowed_names:
					raise NameError(f'Use of "{name}" is not allowed.')

			await ctx.send(eval(code, {"__builtins__": {}}, allowed_names))
		except NameError as e:
			await ctx.send(e)
		except Exception as e:
			await ctx.send(f'Python error: {e}')


	@commands.command(hidden=True)
	@commands.cooldown(3, 15)
	async def reverse(self, ctx, *, message: str):
		'''Reverses a message'''
		await ctx.send(message[::-1])


	@commands.command(hidden=True)
	@commands.cooldown(3, 15)
	async def rot13(self, ctx, *, message: str):
		'''Rotates each letter 13 letters through the alphabet'''
		message = message.lower()
		new_string = ''
		alphabet = 'abcdefghijklmnopqrstuvwxyz'
		for char in message:
			index = alphabet.find(char)
			if index != -1:
				new_index = (index + 13) % 26
				new_string += alphabet[new_index]
			else:
				new_string += char

		await ctx.send(new_string)


def setup(bot):
	bot.add_cog(Other(bot))
